crate-tmp/src/nvtxw.rs

```diff
@@ -0,0 +1,248 @@
+/*
+ *  Copyright (c) 2023-2024, NVIDIA CORPORATION.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  Licensed under the Apache License v2.0 with LLVM Exceptions.
+ *  See LICENSE.txt for license information.
+ *
+ *  SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ */
+
+use std::ffi::c_void;
+use std::ffi::CString;
+use std::mem::size_of;
+use std::path::PathBuf;
+use std::ptr::{null, null_mut};
+
+use crate::nvtxw_bindings;
+
+pub use crate::nvtxw_bindings::nvtxPayloadData_t as PayloadData;
+pub use crate::nvtxw_bindings::nvtxPayloadSchemaAttr_t as PayloadSchemaAttr;
+pub use crate::nvtxw_bindings::nvtxPayloadSchemaEntry_t as PayloadSchemaEntry;
+pub use crate::nvtxw_bindings::nvtxwInterfaceCore_t as InterfaceHandle;
+pub use crate::nvtxw_bindings::nvtxwResultCode_t as ResultCode;
+use crate::nvtxw_bindings::nvtxwSessionAttributes_t as SessionAttributes;
+pub use crate::nvtxw_bindings::nvtxwSessionHandle_t as SessionHandle;
+use crate::nvtxw_bindings::nvtxwStreamAttributes_t as StreamAttributes;
+pub use crate::nvtxw_bindings::nvtxwStreamHandle_t as StreamHandle;
+
+use crate::nvtxw_bindings::NVTXW3_INIT_MODE_LIBRARY_FILENAME;
+use crate::nvtxw_bindings::NVTXW3_INTERFACE_ID_CORE_V1;
+use crate::nvtxw_bindings::NVTXW3_RESULT_FAILED;
+use crate::nvtxw_bindings::NVTXW3_RESULT_INVALID_ARGUMENT;
+use crate::nvtxw_bindings::NVTXW3_RESULT_SUCCESS;
+use crate::nvtxw_bindings::NVTXW3_STREAM_ORDERING_SKID_NONE;
+
+pub const NVTX_PAYLOAD_ENTRY_FLAG_POINTER: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_ENTRY_FLAG_POINTER as u64;
```

Comment:
I implemented this change locally but was discouraged to find that Rust has no implicit conversion from an enum to its underlying integer type, and even moreso that there is no way to convert an integer to an enum of corresponding value other than a `match` block listing every possibility. Adjusting the API so that these are no longer an issue would involve using less and less of bindgen's output directly and instead generating it ourselves or maintaining it manually. I am still investigating a suitable path forward.

Reply:
There are several third-party crates to do this, but in the main compiler we use [`num_enum`](https://docs.rs/num_enum/latest/num_enum/).

Reply:
I've revisited this change with the addition of the `num_enum` crate, which I found to be helpful. Unfortunately, the resulting code is still a bit ugly, because every use of the flag/type enums in the user code must be accompanied by `.into()` to fit into the struct types generated by bindgen, and also because the BitOr trait cannot be applied to enums. Additionally, I encountered this confusing error at the first instance of a flag with `.into()` in the user code, despite the `flags` member and the representation of the `PayloadEntryFlag` enum being `u64`.

```
error[E0283]: type annotations needed
   --> src/nvtxw.rs:248:58
    |
248 |             flags: nvtxw::PayloadEntryFlag::EventMessage.into()
    |                                                          ^^^^
    |
    = note: cannot satisfy `_: std::convert::From<PayloadEntryFlag>`
    = note: required for `PayloadEntryFlag` to implement `Into<_>`
help: try using a fully qualified path to specify the expected types
    |
248 |             flags: <PayloadEntryFlag as Into<T>>::into(nvtxw::PayloadEntryFlag::EventMessage)
    |                    ++++++++++++++++++++++++++++++++++++                                     ~

```

At this point I'm wondering what I should do now. I suspect that the difficulty I'm encountering is because using enums in this way is not very Rust-like. For example, instead of bitwise ORing together flags, a more idiomatic approach might look something more like `std::fs::OpenOptions`. That said, redesigning the API in time for merge is an imposing proposition.

Another idea I had that might offer some degree of improvement is something like this:

```rs
#[allow(non_snake_case)]
pub mod PayloadEntryFlag {
    use crate::nvtxw_bindings;
    pub const POINTER: u64 = nvtxw_bindings::NVTX_PAYLOAD_ENTRY_FLAG_POINTER as u64;
    pub const EVENT_MESSAGE: u64 = nvtxw_bindings::NVTX_PAYLOAD_ENTRY_FLAG_EVENT_MESSAGE as u64;
    // ...
}
```

```rs
        let name_schema = [nvtxw::PayloadSchemaEntry {
            flags: nvtxw::PayloadEntryFlag::EVENT_MESSAGE
                | nvtxw::PayloadEntryFlag::ARRAY_ZERO_TERMINATED,
            // ...
```

Reply:
Let's defer this item for now. I don't want to rush a last-minute API change.

Longer-term, I agree that `std::fs::OpenOptions` is a good example to learn from. We'll still need to play with it to see how much it makes sense. But if you could write an API like, e.g.:

```rust
nvtxw::PayloadEntryFlag::event().array_zero_terminated()
```

Maybe that's better. Certainly, it provides an opportunity for validation which bitwise-or-ing of flags does not provide, and the builder pattern is more Rust-ish in general. The main thing is just the sheer number of flags involved. But maybe that's ok.

Reply:
Thanks for your flexibility. I definitely want to do my research in order to design a robust API.